### PR TITLE
Editor: Add a deprecation dialog

### LINKED ISSUE #41084: Editor: Deprecation notice for WordPress.com editor
> As a user who uses the WordPress.com editor (aka Calypso editor) for a site,
> I need to know that the WordPress.com editor is being deprecated,
> So that I can know how to edit my site content after the deprecation.

## Requirements

1. Users who currently edit a site with the WordPress.com editor see a notice that gives a notification that they will be defaulted to the Block Editor (in Calypso) when the WordPress.com editor is deprecated.
2. Clicking “Try it out” will simply take you to the Gutenberg editor to take it for a spin. No default editor preferences will change.
3. Clicking “Not now” will simply dismiss the modal.
4. Notice is displayed starting ~2 weeks before the WordPress.com editor is deprecated
5. Notice is translated before it's deployed to production

## Notes

[Example notice from core](https://a.cl.ly/d5u0dnDD)

[Example placement for the notice](https://a.cl.ly/DOu85GL0)

### PR DESCRIPTION
**Pull request by pento**
#### Changes proposed in this Pull Request

This change adds a popup deprecation dialog to the Calypso editor, warning that it will soon be replaced by the block editor.

#### Testing instructions

* Visit `/post/mysite.wordpress.com`
* (If you're redirected to the block editor, open the `...` menu and select "Switch to Classic Editor".)
* Confirm that the dialog pops up.
* Confirm that it can be dismissed.
* Confirm that it doesn't display again after being dismissed.
* Confirm that clicking the "Try it out" button switches you to the block editor.
* After being switched to the block editor, confirm that visiting `/post/mysite.wordpress.com` will continue redirecting you to the block editor.

Fixes #41084.

### CODE REVIEW

File: client/post-editor/editor-deprecation-dialog/index.jsx
Comment (a8ci18n) [inline client/post-editor/editor-deprecation-dialog/index.jsx:89]: :information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **6** times:
`translate( 'Try it out: ' )` *ES Score: 8*
See 1 additional suggestions in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/add/41084-editor-deprecation-dialog)

**Comment by jancavan** [comment]
Thanks @pento. I'm afraid I'm not seeing any dialog box. I only see this:

Screen Shot 2020-04-24 at 2 38 06 PM

> (If you're redirected to the block editor, open the ... menu and select "Switch to Classic Editor".)

I also tried switching to the Classic Editor and back, but I'm still not seeing the dialog box.

File: client/state/action-types.js
Comment (scruffian) [inline client/state/action-types.js:222]: I like the convention of these being named as verbs

**Comment by scruffian** [comment]
I noticed a few issues:

This learn more button doesn't seem to work anymore:
Screenshot 2020-04-27 at 12 36 18


In smaller screens the modal looks a bit strange:
Screenshot 2020-04-27 at 12 36 05
Screenshot 2020-04-27 at 12 35 34
Screenshot 2020-04-27 at 12 35 27

File: client/state/ui/editor-deprecation-dialog/actions.js
Comment (scruffian) [inline client/state/ui/editor-deprecation-dialog/actions.js:13]: Why did you choose to use two different functions rather than one and a parameter?

File: client/post-editor/editor-deprecation-dialog/index.jsx
Comment (scruffian) [inline client/post-editor/editor-deprecation-dialog/index.jsx:32]: Since this is a new component maybe it would be better to use a Functional component, so we can take advantage of hooks? On the other hand I guess we'll be deleting this in a few months so it probably doesn't matter much :)

File: client/post-editor/post-editor.jsx
Comment (scruffian) [inline client/post-editor/post-editor.jsx:305]: This change is what breaks the learn more button. I think we leave this line how it is.

**Comment by scruffian** [comment]
The EditorGutenbergBlocksWarningDialog shows when a post with blocks is loaded in the classic editor. I wonder if we should use the same check to not show this new dialog.

File: client/post-editor/editor-deprecation-dialog/index.jsx
Comment (scruffian) [inline client/post-editor/editor-deprecation-dialog/index.jsx:64]: We also have a <Dialog> component. I'm not sure why we have two, or which one to use...

File: client/state/ui/editor-deprecation-dialog/actions.js
Comment (scruffian) [inline client/state/ui/editor-deprecation-dialog/actions.js:6]: It doesn't look like this is ever called...

File: client/post-editor/editor-deprecation-dialog/index.jsx
Comment (scruffian) [reply client/post-editor/editor-deprecation-dialog/index.jsx:64]: Oh I see `Modal` comes from core, so that's probably the better choice :)

**Comment by scruffian** [comment]
We already have a few different dialogs on the editor. I added a [commit](https://github.com/Automattic/wp-calypso/pull/41391/commits/d1aa472ebf6fc29d9f268de77e6723a0935f00bd) to make sure that we are still able to show the other dialogs if needed, but prevent us showing them all at the same time.

To test this you can force the warning dialog to show by changing `isDialogVisible` on this line to `true`:
https://github.com/Automattic/wp-calypso/blob/4537033c6794449ce0a10b69d72daf5fa658d553/client/post-editor/editor-gutenberg-blocks-warning-dialog/index.jsx#L75

When you do this you'll see this without my commit above:
Screenshot 2020-04-28 at 14 07 22

And this with it:
Screenshot 2020-04-28 at 14 57 32
Screenshot 2020-04-28 at 14 57 11

**Comment by scruffian** [comment]
I raised some questions about the way this should work here: pb5gDS-uw-p2

**Comment by davemart-in** [comment]
> This is kind of obnoxious as it means in some circumstances we are showing two dialogs, but the number of cases of this will probably be small

I agree. I'm okay with having this as an edge case situation. The number of people who will see this should be small since the modal only shows once.

**Comment by scruffian** [comment]
I've updated the requirements on #41084 based on the post...

File: client/state/ui/editor-deprecation-dialog/actions.js
Comment (pento) [reply client/state/ui/editor-deprecation-dialog/actions.js:6]: Good point, removed.

File: client/state/ui/editor-deprecation-dialog/actions.js
Comment (pento) [reply client/state/ui/editor-deprecation-dialog/actions.js:6]: Copy/pasta from a different action, mostly. 🙂 I've deleted the one that isn't being used.

File: client/post-editor/editor-deprecation-dialog/index.jsx
Comment (pento) [reply client/post-editor/editor-deprecation-dialog/index.jsx:34]: Eh, I'm not so concerned about the amazing-ness of code that's born to die. 😉

**Comment by pento** [comment]
Thanks for trying to test, @jancavan. Were you trying it on calypso.live? If so, I've now set it so that it shows there.

Thank you for the feedback and fix, @scruffian. I've implemented fixes for all of the problems you've pointed out. It's worth noting that there's a little bit of jankiness on some screen sizes around the CSS breakpoints, primarily due to WordPress components having a 600px breakpoint, and Calypso having a 660px breakpoint.

**Comment by jancavan** [comment]
Thanks @pento. I can see it now.

There's discussion here about changing the button copy: p2-pb5gDS-uw-#comment-1283

Entire copy is still going to be reviewed, but we'll want to change "Read more here" to just say "Read more"
Screen Shot 2020-04-30 at 11 49 28 AM

It looks like there's a lot of empty blue space above the illustration, it didn't look like that in the mockups. cc @dwolfe 

Is it supposed to take up the entire screen on mobile?

**Comment by dwolfe** [comment]
Updated modal copy given @glendaviesnz's [comment here](https://github.com/Automattic/wp-calypso/pull/41570#issuecomment-621523371):

Pre-sunset Calypso Editor modal

@jancavan It looks like the illustration in your screenshot (185px square) is smaller than it is in our comps (250px square). Also, it doesn't display at all on mobile, based on the existing welcome modal:

IMG_4613

**Comment by jancavan** [comment]
>  It looks like the illustration in your screenshot (185px square) is smaller than it is in our comps

Yes, the screenshot I took was from this PR:
https://calypso.live/?branch=add/41084-editor-deprecation-dialog

**Comment by pento** [comment]
@jancavan: I need a better copy of the illustration, I don't believe the one currently in this PR is correct. I'll be able to adjust the size to be more accurate if the illustration doesn't have borders.



It's also worth noting that, as the mockups are using WordPress core styling, I've used the WordPress `Modal` component, which probably behaves differently to Calypso's `Dialog` component under some circumstances (for example, on mobile).

**Comment by pento** [comment]
@dwolfe: It seems that the copy changes are a little premature, since this PR implements a temporary switch.

However, there currently isn't an indication within the block editor that this is a temporary switch, with the easy ability to return to the classic editor, or make the change permanent. Perhaps the sidebar notice in #41087 could include the ability to make the change permanent?

**Comment by pento** [comment]
I've changed the dialog back to switching the user's editor choice, rather than the block editor trial functionality.

Note that this PR requires a better copy of the illustration.

**Comment by scruffian** [comment]
> It's worth noting that there's a little bit of jankiness on some screen sizes around the CSS breakpoints, primarily due to WordPress components having a 600px breakpoint, and Calypso having a 660px breakpoint.

Yeah I saw this too. We should align them...

File: client/post-editor/editor-deprecation-dialog/style.scss
Comment (scruffian) [inline client/post-editor/editor-deprecation-dialog/style.scss:9]: If we use the `$break-small` mixin from Gutenberg we can use the same breakpoints as Gutenberg

**Comment by pento** [comment]
@jancavan, @dwolfe, @davemart-in: Where can I find a copy of the illustration I need to use here?

**Comment by davemart-in** [comment]
@dwolfe do you have a vector of that image for @pento somewhere?

File: client/post-editor/editor-deprecation-dialog/style.scss
Comment (pento) [reply client/post-editor/editor-deprecation-dialog/style.scss]: Good idea, done in 50a7122a07.

**Comment by pento** [comment]
This PR is just waiting on the correct illustration.

It's worth noting that the current behaviour is a little weird on screens just slightly wider than 600px, but how it changes will depend on how the illustration fits, so I'll leave it until I have the correct file to work with.

**Comment by scruffian** [comment]
The changes look good. I agree that this will be good to merge once we have the illustration.

**Comment by dwolfe** [comment]
I couldn't find that vector in anything but screenshot form, but I made one with the same dimensions as one of @jancavan's comps, and updated with the icons that are in production:

[Blocks_Illustration.svg.zip](https://github.com/Automattic/wp-calypso/files/4582252/Blocks_Illustration.svg.zip)

Let me know if this isn't what you need.

**Comment by dwolfe** [comment]
>@dwolfe do you have a vector of that image for @pento somewhere?

Couldn't find it anywhere, so I worked with @jancavan to whip one up. This has updated icons from the actual Gutenberg blocks live in production right now, and the grey block backgrounds match the other graphics that Jan provided for the Customer Home announcement.

[Blocks_Illustration_v2.svg.zip](https://github.com/Automattic/wp-calypso/files/4582574/Blocks_Illustration_v2.svg.zip)

**Comment by pento** [comment]
Thank you for the updated image, @jancavan and @dwolfe, it works perfectly! I've tweaked the layout to more closely match the mockups, here are some current screenshots:

File: client/post-editor/editor-deprecation-dialog/index.jsx
Comment (a8ci18n) [inline client/post-editor/editor-deprecation-dialog/index.jsx:79]: :information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **3** times:
`translate( 'Get a head start before we activate it for everyone on {{date/}}. {{support}}Read more{{/support}}.' )` *ES Score: 8*

**Comment by Stojdza** [comment]
@pento e2e tests have passed with [this](https://github.com/Automattic/wp-calypso/pull/41391/commits/f71ae978ae2a03e4c1875b55d75e42a81d582768) change.

**Review by scruffian — APPROVED** [review]
This looks good to me. The design might want another pair of eyes..

**Comment by dwolfe** [comment]
Hey @pento, sorry to add another step to this, but the second image in [your comment above](https://github.com/Automattic/wp-calypso/pull/41391#issuecomment-624401329) - is that the responsive modal at narrow screen sizes?

If so, can we keep the blue background, and add a 24px margin above the headline, like this?

Responsive-480px-and-below

Also, as @jancavan were discussing just now, we realized there might be a gap between the width of the desktop modal (750px) and mobile screen sizes (say, 480px) where having the modal cover the viewport would generate a lot of negative space. Between those breakpoints, could we keep the modal at 480px wide, like this?

Responsive, _ 480px

**Comment by pento** [comment]
Thanks for the feedback, @dwolfe! I've implemented the background colour and margin in 95f7ab2.

Unfortunately, I don't think there's a practical way to implement that medium sized modal: the WordPress Modal component [only breaks at 600px](https://github.com/WordPress/gutenberg/blob/2beb59861039c43c71f03c25256479b30c148387/packages/components/src/modal/style.scss#L16), it doesn't have another breakpoint we could use. I could potentially implement something between 600px and 782px (the next breakpoint up in WordPress), but below 600px, the fullscreen state of the modal is unavoidable.

Here's how it currently looks on mobile:

**Comment by dwolfe** [comment]
I see. Well, sounds like it is what it is, although it's not as awkward as I imagined, looking at that screenshot. One last little thing: can we make the margin between the top of the screen and the illustration 16px?

After that it's 👍 from me. Thanks so much for all your work on this, @pento!

**Pull request merged**
